src/backend/oneapi/solve.cpp

```diff
@@ -127,71 +168,81 @@ Array<T> leastSquares(const Array<T> &a, const Array<T> &b) {
 
         // QR is performed on the transpose of A
         Array<T> A = transpose<T>(a, true);
-
-#if UNMQR
-        const dim4 NullShape(0, 0, 0, 0);
         dim4 endPadding(N - b.dims()[0], K - b.dims()[1], 0, 0);
         B = (endPadding == NullShape
                  ? copyArray(b)
                  : padArrayBorders(b, NullShape, endPadding, AF_PAD_ZERO));
-        B.resetDims(dim4(M, K));
-#else
-        B = copyArray<T>(b);
-#endif
-
-        int NB       = magma_get_geqrf_nb<T>(A.dims()[1]);
-        int NUM      = (2 * MN + ((M + 31) / 32) * 32) * NB;
-        Array<T> tmp = createEmptyArray<T>(dim4(NUM));
 
-        vector<T> h_tau(MN);
+        // Get workspace needed for QR
+        std::int64_t scratchpad_size =
+            ::oneapi::mkl::lapack::geqrf_scratchpad_size<compute_t<T>>(
+                getQueue(), A.dims()[0], A.dims()[1], A.strides()[1]);
 
-        int info   = 0;
-        Buffer *dA = A.get();
-        Buffer *dT = tmp.get();
-        Buffer *dB = B.get();
+        sycl::buffer<compute_t<T>> scratchpad(scratchpad_size);
+        Array<compute_t<T>> t = createEmptyArray<T>(af::dim4(MN, 1, 1, 1));
 
-        magma_geqrf3_gpu<T>(A.dims()[0], A.dims()[1], (*dA)(), A.getOffset(),
-                            A.strides()[1], &h_tau[0], (*dT)(), tmp.getOffset(),
-                            getQueue()(), &info);
+        sycl::buffer<compute_t<T>> aBuf =
+            A.template getBufferWithOffset<compute_t<T>>();
+        sycl::buffer<compute_t<T>> tBuf =
+            t.template getBufferWithOffset<compute_t<T>>();
+        // In place Perform in place QR
+        ::oneapi::mkl::lapack::geqrf(getQueue(), A.dims()[0], A.dims()[1], aBuf,
+                                     A.strides()[1], tBuf, scratchpad,
+                                     scratchpad_size);
 
+        // R1 = R(seq(M), seq(M));
         A.resetDims(dim4(M, M));
 
-        magmablas_swapdblk<T>(MN - 1, NB, (*dA)(), A.getOffset(),
-                              A.strides()[1], 1, (*dT)(),
-                              tmp.getOffset() + MN * NB, NB, 0, queue);
+        // Bt = tri_solve(R1, B);
+        B.resetDims(dim4(M, K));
 
-        OPENCL_BLAS_CHECK(
-            gpu_blas_trsm(OPENCL_BLAS_SIDE_LEFT, OPENCL_BLAS_TRIANGLE_UPPER,
-                          OPENCL_BLAS_CONJ_TRANS, OPENCL_BLAS_NON_UNIT_DIAGONAL,
-                          B.dims()[0], B.dims()[1], scalar<T>(1), (*dA)(),
-                          A.getOffset(), A.strides()[1], (*dB)(), B.getOffset(),
-                          B.strides()[1], 1, &queue, 0, nullptr, &event));
+        sycl::buffer<compute_t<T>> bBuf =
+            B.template getBufferWithOffset<compute_t<T>>();
```

Comment:
Maybe we should make getBufferWithOffset as a free function? I don't really like the usage here but its your call.